### ISSUE TITLE
Add Interactive Highlights to GrabDragInteraction

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/scenery-phet#698]
For https://github.com/phetsims/balloons-and-static-electricity/issues/516 and https://github.com/phetsims/friction/issues/200 we need to add support to GrabDragInteraction to have unique highlights for "Interactive Highlights" in the grab and drag state. The focus highlights cannot be used because children are added to these highlights in GrabDragInteraction that have content that is specific to alternative input that should be displayed when the target Node has focus.

**Comment by zepumph** [comment]
I'm seeing some really weird stuff over in friction, where the solid highlight around the Friction books are appearing even when interactive highlights are not enabled, and are sticking around even when I click on something else. 

It seems like this is a work in progress, and you are not blocking me, but I thought I'd note in case you weren't aware of the bug.

**Comment by jessegreenberg** [comment]
Thanks! I was not aware of this. This was not ready for review until some renaming was done but I had thought was otherwise stable. Will take a look.

**Comment by jessegreenberg** [comment]
I was able to get this to happen once in Friction, but have not been able to again so I can't find the case that is causing this. @zepumph are you able to get this to happen consistently, or do you know the steps to get it to happen by chance?

**Comment by jessegreenberg** [comment]
I was able to get it to happen once by
1) Enabling Interactive Highlights
2) Grabbing the zoomed in book
3) Tabbing to the zoomed out book while mouse is down on the zoomed in book
4) Grabbing in the zoomed out book with "enter key", without releasing mouse.

In this case I saw a "grabbed" dotted highlight around both the zoomed in and zoomed out books. But trying to reproduce by following these steps and it isn't happening again. So something else must have put it in this state to get it to happen.

**Comment by jessegreenberg** [comment]
I think I got it!

1) Enable Interactive Highlights
2) Mouse click down on zoomed in chemistry book
3) Move your mouse over the zoomed out chemistry book, while dragging the zoomed in chemistry book
4) While holding mouse down, tab to the zoomed out chemistry book, and grab it by pressing enter
5) Move the pressed mouse back to the zoomed in chemistry book. Then release it.

**Comment by jessegreenberg** [comment]
In a meeting today, @zepumph pointed out that just turning on Voicing caused interactive highlights to appear in Friction.

**Comment by jessegreenberg** [comment]
This bug was fixed by changing two things.

First, when changing the Interactive Highlight in InteractiveHighlight.js it was important to set the highlight invisible so that the HighlightOverlay is responsible for displaying it. Always setting it visible whenever the Node is "activated" by a Pointer missed important logic in HighlightOverlay.

Second, the check to prevent activating the Interactive Highlights when PDOM focus highlights are visible was not quite in the right place. It has been moved so that Interactive Highlights should never appear while there is DOM focus.

**Comment by jessegreenberg** [comment]
I don't anticipate any more changes here, even though https://github.com/phetsims/scenery/issues/1230 has a chance to rename some things. But I think this is ready for review.

@zepumph would you mind reviewing the changes to GrabDragInteraction. Changes can be found by searching for the added `grabInteractiveHighlight` and `dragInteractiveHighlight`.  The motivation for these changes is described by this added comment

```js
    // [...] The Interactive Highlights cannot
    // fall back to the default focus highlights because GrabDragInteraction adds "grab cue" Nodes as children
    // to the focus highlights that should not be displayed when using Interactive Highlights.
```

This is making use of a new feature added in https://github.com/phetsims/scenery/issues/1274, where you can set the Interactive highlights separately from focus highlight so they can be different.

**Comment by zepumph** [comment]
I general I feel not very up-to-date about these changes. I have a lot of questions about interactive highlights that may not apply to this review, but this is where I encountered them. Let me know if we should discuss in person instead of async in this issue.

>https://github.com/phetsims/scenery-phet/blob/94f3080a46d8bd8f6500e833113d8adb7734eaba/js/accessibility/GrabDragInteraction.js#L186-L188

- [ ] Can't we just use the focusHighlight by default? Is it already doing this without me realizing?

>https://github.com/phetsims/friction/blob/f06d3c36a5f1d977bcf9fc9871aef68863c44f4e/js/friction/view/book/BookNode.js#L133-L135

- [ ] Why do we specify this for focusHighlight, but not for interactiveHighlights?

>https://github.com/phetsims/scenery-phet/blob/94f3080a46d8bd8f6500e833113d8adb7734eaba/js/accessibility/GrabDragInteraction.js#L195-L198

Should this check on `node.interactiveHighlight` instead?

>https://github.com/phetsims/scenery-phet/blob/94f3080a46d8bd8f6500e833113d8adb7734eaba/js/accessibility/GrabDragInteraction.js#L357

- [ ] This feels incomplete to me, why are we using the transformSourceNode from the focusHighlight on the interactiveHighlight? First off we should make sure that the transformSourceNode of grabInteractiveHighlight is transferred, and not this.grabFocusHighlight.
- [ ] Second: There are many other options on FocusHighlightNode that are not transferred to the drag highlight. Presumably we would want to support all of these right? And then just make it dashed? https://github.com/phetsims/scenery/blob/017ec43ea433b606c1874424eb14eeac1bb9ebc5/js/accessibility/FocusHighlightPath.js#L50-L69

>https://github.com/phetsims/scenery-phet/blob/94f3080a46d8bd8f6500e833113d8adb7734eaba/js/accessibility/GrabDragInteraction.js#L368-L372

Why is there no parallel implementation here for focusHighlight? Perhaps we should move this to where focusHighlightLayerable logic is checked. Like perhaps here? https://github.com/phetsims/scenery-phet/blob/94f3080a46d8bd8f6500e833113d8adb7734eaba/js/accessibility/GrabDragInteraction.js#L422-L433. Otherwise please document why these two implementation for two different highlights have deviated on purpose. 

>https://github.com/phetsims/scenery-phet/blob/94f3080a46d8bd8f6500e833113d8adb7734eaba/js/accessibility/GrabDragInteraction.js#L385-L390

- [ ] I understand that we just duplicated everything above such that we didn't have to add this logic to the interactive highlight, but could you please document that specifically here, like "We want this for focus highlight but not for interactive highlight because. . . "

- [ ] Is it worth duplicating this assertion for interactive highlights? https://github.com/phetsims/scenery-phet/blob/94f3080a46d8bd8f6500e833113d8adb7734eaba/js/accessibility/GrabDragInteraction.js#L552-L554

- [ ] In this and the usage right below, why are we not toggling visibility on interactiveHighlights? If this is correct as it currently is, please document.

https://github.com/phetsims/scenery-phet/blob/94f3080a46d8bd8f6500e833113d8adb7734eaba/js/accessibility/GrabDragInteraction.js#L29-L34

- [ ] If this applies to interactiveHighlight please update accordingly.


Thanks for all the hard work on this! Let me know what I can do to help.

**Comment by zepumph** [comment]
- [ ] Was BASE RC shas taken supporting Interactive Highlights? If so I would be worried about its implementation without addressing some of these concerns, but let's talk about it.

**Comment by jessegreenberg** [comment]
Thank you very much for the thorough review! Responding to each:
> Can't we just use the focusHighlight by default? Is it already doing this without me realizing?

Often we can, but not for GrabDragInteraction because of https://github.com/phetsims/scenery-phet/blob/94f3080a46d8bd8f6500e833113d8adb7734eaba/js/accessibility/GrabDragInteraction.js#L346-L348. And so if `interactiveHighlightLayerable` is true it needs to be added to the scene graph on construction like the focus highlights.

> Why do we specify this for focusHighlight, but not for interactiveHighlights?

Good catch, the assertion was incorrect, it should be the `interactiveHighlight`. Fixed in https://github.com/phetsims/scenery-phet/commit/23eac632c7fec7208f894e270ae7bfa39f7eea07

> First off we should make sure that the transformSourceNode of grabInteractiveHighlight is transferred, and not this.grabFocusHighlight.

Yes, you are correct! It it should use the `grabInteractiveHighlight` instead. Done in https://github.com/phetsims/scenery-phet/commit/abfccf1f66183c700ada5ea1d8bdac38b4729d22

> here are many other options on FocusHighlightNode that are not transferred to the drag highlight. Presumably we would want to support all of these right? And then just make it dashed?

Potentially, though I could also see an argument that the look and feel for `dragFocusHighlight` and `dragInteractiveHighlight` should look the same for all things using GrabDragInteraction. If you still feel like this would be an improvement can you please make a new issue to add support for more customizable drag highlights?

> Why is there no parallel implementation here for focusHighlight? Perhaps we should move this to where focusHighlightLayerable logic is checked. Like perhaps here?

The logic for `focusHighlightLayerable` is checked in the click listener which likely will never be triggered when using Interactive Highlights. Alternatively, could the logic for focus highlights be moved to the constructor where it is done for Interactive Highlights? Don't these assertions mean that would be safe? https://github.com/phetsims/scenery-phet/blob/abfccf1f66183c700ada5ea1d8bdac38b4729d22/js/accessibility/GrabDragInteraction.js#L179-L185

> I understand that we just duplicated everything above such that we didn't have to add this logic to the interactive highlight, but could you please document that specifically here, like 

Done, I think I got the sentiment of the comment in https://github.com/phetsims/scenery-phet/commit/406d2484ba2ef931c70ee079c2052e91a7472010

>Is it worth duplicating this assertion for interactive highlights?

That makes sense, I added it in https://github.com/phetsims/scenery-phet/commit/a616b2951247e22364c87da444c6c305a3e456b6

> In this and the usage right below, why are we not toggling visibility on interactiveHighlights? If this is correct as it currently is, please document.

I added the matching lines for Interactive Highlights. But I also think these should be not necessary for both focus and interactive highlights. The HighlightOverlay should update visibility for us. Would you be OK if I tried removing them?

> If this applies to interactiveHighlight please update accordingly.

Thanks, same conditions apply and documentation updated in https://github.com/phetsims/scenery-phet/commit/388ceca27a430642b2fa9f709e2afe676092be88

> Was BASE RC shas taken supporting Interactive Highlights? If so I would be worried about its implementation without addressing some of these concerns, but let's talk about it.

Yea...I probably jumped the gun to recommend BASE SHAs. But I am not too concerned yet. Problems identified so far mostly involve `interactiveHighlightsLayerable` or `transformSourceNode` and BASE use neither. BASE also went through a dev test with these changes before taking RC SHAs.

OK back to @zepumph to review comments.

**Comment by zepumph** [comment]
>The logic for focusHighlightLayerable is checked in the click listener which likely will never be triggered when using Interactive Highlights. Alternatively, could the logic for focus highlights be moved to the constructor where it is done for Interactive Highlights? Don't these assertions mean that would be safe?

I think we should just leave it. I would rather not mess with focusHighlightLayerable logic. It has been working for quite some time, and its quite complicated.

>I added the matching lines for Interactive Highlights. But I also think these should be not necessary for both focus and interactive highlights. The HighlightOverlay should update visibility for us. Would you be OK if I tried removing them?


Yes please. I trust you, and know that you are the person to do it and test that it is working as expected. Good luck!

**Comment by zepumph** [comment]
Thanks @jessegreenberg, I am nervous to hear about BASE, but I'm also out of the loop about the entire feature and implementation. I defer to your judgement. Thanks for all this work! Especially since it has been so timely from my review earlier today.

**Comment by zepumph** [comment]
I don't think there is anything else left for this. @jessegreenberg?
